File: Python3/lib/operators.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Fri Jan 20 07:00:46 2017

@author: jedi
"""

import math
from lib import lex, execute

class Number:
    def __init__(self, lexi, denominator=None):
        if denominator is None:
            self.numerator = float(lexi.data)
            self.denominator = 1
            self.reduce()
        else:
            self.numerator = lexi
            self.denominator = denominator
            self.reduce()
            
    def reduce(self):
        if self.numerator == 0:
            if self.denominator != 0:
                self.denominator = 1
            return self
        while self.numerator % 1 != 0.0:
            self.denominator *= 10
            self.numerator *= 10
        while self.denominator % 1 != 0.0:
            self.denominator *= 10
            self.numerator *= 10
        gcd = math.gcd(int(self.denominator), int(self.numerator))
        self.denominator /= gcd
        self.numerator /= gcd
        
        if self.denominator < 0:
            self.numerator = -self.numerator
            self.denominator = -self.denominator
        return self
    
    def toInt(self):
        return int(self.numerator / self.denominator)
    
    def toFloat(self):
        return self.numerator / self.denominator
    
    def __add__(self, num):
        gcd = int(self.denominator * num.denominator / math.gcd(int(self.denominator), int(num.denominator)))
        return Number(self.numerator * gcd / self.denominator + num.numerator * gcd / num.denominator, gcd).reduce()
    
    def __sub__(self, num):
        gcd = int(self.denominator * num.denominator / math.gcd(int(self.denominator), int(num.denominator)))
        return Number(self.numerator * gcd / self.denominator - num.numerator * gcd / num.denominator, gcd).reduce()
    
    def __mul__(self, num):
        return Number(self.numerator * num.numerator, self.denominator * num.denominator).reduce()
        
    def __truediv__(self, num):
        return Number(self.numerator * num.denominator, self.denominator * num.numerator).reduce()
    
    def __float__(self):
        return self.numerator * 1.0 / self.denominator
    
    def __eq__(self, num):
        return self.denominator == num.denominator and self.numerator == num.numerator
    
    def __gt__(self, num):
        return (self - num).numerator > 0
    
    def __ge__(self, num):
        return (self - num).numerator >= 0
    
    def __lt__(self, num):
        return (self - num).numerator < 0

    def __le__(self, num):
        return (self - num).numerator <= 0
        
    def __repr__(self):
        return str(int(self.numerator)) + "/" + str(int(self.denominator)) if self.denominator != 1 else str(int(self.numerator))

class Set:
    def __init__(self, first_val=None):
        if first_val == None:
            self.data = []
        elif type(first_val) == list:
            self.data = first_val
#        elif type(first_val) != Number:
#            first_val = Number(first_val)
#            self.data = [first_val]
        elif type(first_val) == Number:
            self.data = [first_val]
    
    def append(self, val):
        if type(val) != Number:
            val = Number(val)
        self.data.append(val)
        self.data.sort()
        
    def __repr__(self):
        return repr(self.data)
    
    def __str__(self):
        return str(self.data)

class CSet:
    pass

class Vector:
    def __init__(self, first_val=None):
        if first_val == None:
            self.data = []
        elif type(first_val) == list:
            self.data = list
        elif type(first_val) != Number:
            first_val = Number(first_val)
            self.data = [first_val]
    
    def append(self, val):
        if type(val) != Number:
            val = Number(val)
        self.data.append(val)

class TVector:
    pass

class Function:
    def __init__(self, var, expr):
        self.var = var
        self.expr = expr
    
    def __repr__(self):
        return 'Vars: ' + str(self.var) + ' Exprs: ' + str(self.expr)
        
    def func(self, vals):
        for i in range(len(self.var)):
            execute.add_var(self.var[i], vals[i])
        ret_val = execute.evaluate(self.expr)
        for i in range(len(self.var)):
            execute.del_var(self.var[i])
        return ret_val

# class String(str):
#     pass

class Boolean(Number):
    def isTrue(self):
        return self.denominator != 0
    
true = Number(1,1)
false = Number(0,0)
        
def varExchange(lexi):
    if type(lexi) == lex.lex_class:
        if lexi.fsm == lex.lex_fsm.NUMBER:
            return Number(lexi)
        elif lexi.fsm == lex.lex_fsm.EXPR:
            return lexi.data
        else:
            raise TypeError(lexi)
    else:
        return lexi

def operate(op, a, b):
    if op == '+':
        return a + b
    if op == '-':
        return a - b
    if op == '*':
        return a * b
    if op == '/':
        return a / b
    if op == '<':
        return true if a < b else false
    if op == '<=':
        return true if a <= b else false
    if op == '>':
        return true if a > b else false
    if op == '>=':
        return true if a >= b else false
    if op == '==':
        return true if a == b else false
    if op == '!=':
        return true if a != b else false
    if op == '\|\|':
        return true if (a == true) or (b == true) else false
    if op == '&&':
        return true if (a == true) and (b == true) else false

File: Go/operators.go
package main

func gcd(a, b int) int {
	for b > 0 {
		a, b = b, a%b
	}
	return a
}

func (self *number) reduce() {
	if self.numerator == 0 {
		if self.denominator != 0 {
			self.denominator = 1
		}
		return
	}
	g := gcd(self.denominator, self.numerator)
	self.denominator /= g
	self.numerator /= g
	if self.denominator < 0 {
		self.numerator = -self.numerator
		self.denominator = -self.denominator
	}
}

func equal(n, n_ number) bool {
	return n.numerator == n_.numerator && n.denominator == n_.denominator
}

func lt(n, n_ number) bool {
	return n.numerator*1.0/n.denominator < n_.numerator*1.0/n.denominator
}

func le(n, n_ number) bool {
	return n.numerator*1.0/n.denominator <= n_.numerator*1.0/n.denominator
}

func gt(n, n_ number) bool {
	return n.numerator*1.0/n.denominator > n_.numerator*1.0/n.denominator
}

func ge(n, n_ number) bool {
	return n.numerator*1.0/n.denominator >= n_.numerator*1.0/n.denominator
}

func add(n, n_ number) (res number) {
	g := n.denominator * n_.denominator / gcd(n.denominator, n_.denominator)
	res = number{n.numerator*g/n.denominator + n_.numerator*g/n_.denominator, g}
	res.reduce()
	return
}

func sub(n, n_ number) (res number) {
	g := n.denominator * n_.denominator / gcd(n.denominator, n_.denominator)
	res = number{n.numerator*g/n.denominator - n_.numerator*g/n_.denominator, g}
	res.reduce()
	return
}

func mul(n, n_ number) (res number) {
	res = number{n.numerator * n_.numerator, n.denominator * n_.denominator}
	res.reduce()
	return
}

func div(n, n_ number) (res number) {
	res = number{n.numerator * n_.denominator, n.denominator * n_.numerator}
	res.reduce()
	return
}

func do_func(lambda storage, argc []storage, variable *Variable) *storage {
	if lambda.vartype == VAR_FUNCTION {
		f := lambda.data.(function)
		return f.function(argc, variable)
	} else if lambda.vartype == VAR_C_FUNCTION {
		f := lambda.data.(inherit)
		return f.function(argc, variable)
	} else {
		panic("not a function")
	}
}
